rsa_test.go

```diff
@@ -13,11 +13,73 @@ import (
 )
 
 func TestRSAKeyGeneration(t *testing.T) {
+	for _, size := range []int{2048, 3072} {
+		size := size
+		t.Run(strconv.Itoa(size), func(t *testing.T) {
+			t.Parallel()
+			_, _, _, _, _, _, _, _, err := openssl.GenerateKeyRSA(size)
+			if err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
+func testRSAEncryptDecryptPKCS1(t *testing.T, priv *openssl.PrivateKeyRSA, pub *openssl.PublicKeyRSA) {
+	msg := []byte("hi!")
+	enc, err := openssl.EncryptRSAPKCS1(pub, msg)
+	if err != nil {
+		t.Fatalf("EncryptPKCS1v15: %v", err)
+	}
+	dec, err := openssl.DecryptRSAPKCS1(priv, enc)
+	if err != nil {
+		t.Fatalf("DecryptPKCS1v15: %v", err)
+	}
+	if !bytes.Equal(dec, msg) {
+		t.Fatalf("got:%x want:%x", dec, msg)
+	}
+}
+
+func TestRSAEncryptDecryptPKCS1(t *testing.T) {
 	for _, size := range []int{2048, 3072} {
 		size := size
 		t.Run(strconv.Itoa(size), func(t *testing.T) {
 			t.Parallel()
 			priv, pub := newRSAKey(t, size)
+			testRSAEncryptDecryptPKCS1(t, priv, pub)
+		})
+	}
+}
+
+func TestRSAEncryptDecryptPKCS1_MissingPrecomputedValues(t *testing.T) {
+	N, E, D, P, Q, Dp, Dq, Qinv, err := openssl.GenerateKeyRSA(2048)
+	if err != nil {
+		t.Fatalf("GenerateKeyRSA: %v", err)
+	}
+	tt := []struct {
+		name    string
+		wantErr bool
+		fn      func() (*openssl.PrivateKeyRSA, *openssl.PublicKeyRSA)
+	}{
+		{"noDp", false, func() (*openssl.PrivateKeyRSA, *openssl.PublicKeyRSA) {
+			return newRSAKeyFromParams(t, N, E, D, P, Q, nil, Dq, Qinv)
+		}},
+		{"noDq", false, func() (*openssl.PrivateKeyRSA, *openssl.PublicKeyRSA) {
+			return newRSAKeyFromParams(t, N, E, D, P, Q, Dp, nil, Qinv)
+		}},
+		{"noQinv", false, func() (*openssl.PrivateKeyRSA, *openssl.PublicKeyRSA) {
+			return newRSAKeyFromParams(t, N, E, D, P, Q, Dp, Dq, nil)
+		}},
+		{"none", false, func() (*openssl.PrivateKeyRSA, *openssl.PublicKeyRSA) {
+			return newRSAKeyFromParams(t, N, E, D, P, Q, nil, nil, nil)
```

Comment:
Maybe it's a bit silly because we know the internal logic is basically just `Dp && Dq && Qinv`, but this could be done exhaustively. (Iterate from `0b001` to `0b111` and skip passing the `1` bits, or something like that.)

Reply:
Doing an exhaustive test seems good, will do that. Won't use your bit test suggestion, though, it is too clever, I'll just hardcode the possible combinations 😄.

Reply:
Another way without as much cleverness (and with some nesting) would be:

```go
	falseTrue := []bool{true, false}
	for _, withDp := range falseTrue {
		for _, withDq := range falseTrue {
			for _, withQinv := range falseTrue {
				//...
			}
		}
	}
```

But still not quite as simple as table-driven possibilities, to be sure!